src/Models/Services/OatUtils.ts

```diff
@@ -135,67 +77,173 @@ export const getDirectoryPathFromDTMI = (dtmi: string) => {
+}
+
+/** does some cleanup on the entities to make them properly shaped for DTDL since we need some extra stuff to manage the lifecycle within the app */
+export function convertModelToDtdl(model: DtdlInterface): DtdlInterface {
+    const newModel = deepCopy(model);
+    newModel.contents?.forEach((x) => {
+        if (x['@type'] !== OAT_INTERFACE_TYPE) {
+            delete x['@id'];
         }
+    });
+    // console.log(`***Converted ${model['@id']}. {model}`, newModel);
+    return newModel;
+}
+
+/** returns the id for the node of an untargeted relationship */
+export function getUntargetedRelationshipNodeId(
+    sourceModelId: string,
+    relationship: DtdlInterfaceContent
+): string {
+    const id =
+        relationship['@id'] || // use the given id if present
+        `${OAT_UNTARGETED_RELATIONSHIP_ID_PREFIX}_${sourceModelId}_${relationship.name}`; // generate a name from the relationship name
+    return id;
+}
+
+/** looks at the id of a given entity and returns whether that is an untargeted entity */
+export function isUntargeted(id: string) {
+    return id?.startsWith(OAT_UNTARGETED_RELATIONSHIP_ID_PREFIX);
+}
+
+//#region Model ID
+
+const DEFAULT_VERSION_NUMBER = 1;
+interface IBuildModelIdArgs {
+    /** namespace for the current ontology */
+    namespace: string;
+    /** name of the model */
+    modelName: string;
+    /** the sub path for the model (optional) */
+    path?: string;
+    /** version number for the model. If omitted, will use the default value */
+    version?: number;
+}
+/**
+ * builds out the version id string for a model
+ * @returns string for the id of the model
+ */
+export function buildModelId({
+    namespace,
+    modelName,
+    path,
+    version
+}: IBuildModelIdArgs): string {
+    const prefix = OAT_MODEL_ID_PREFIX;
+    const namespaceValue = namespace?.replace(/ /g, '');
+    const pathValue = path?.replace(/ /g, '');
+    const nameValue = modelName?.replace(/ /g, '');
+    const versionNumber = isDefined(version) ? version : DEFAULT_VERSION_NUMBER;
+
+    let uniqueName = nameValue;
+    if (pathValue) {
+        uniqueName = pathValue + ':' + nameValue;
     }
-
-    return modelsCopy;
-};
+    return `${prefix}:${namespaceValue}:${uniqueName};${versionNumber}`;
+}
+
+export function parseModelId(
+    id: string
+): {
+    namespace: string;
+    name: string;
+    path: string;
+    version: string;
+} {
+    if (!id) {
+        return {
+            name: '',
+            namespace: '',
+            path: '',
+            version: ''
+        };
+    }
+    const getNamespace = (id: string) => {
+        return id.substring(0, id.indexOf(':'));
+    };
+    const getPath = (id: string) => {
+        // if we still have any : then they must be part of the path or the separator
+        if (id.split(':').length > 0) {
+            return id.substring(0, id.lastIndexOf(':'));
+        }
+        return '';
+    };
+    const getName = (id: string) => {
```

Comment:
I don't see any input validations if there are malformed dtmi's with multiple segments with `;`

Reply:
Similar, empty/whitespace

Reply:
I'm thinking there should be a call to something like https://github.com/Azure/iot-plugandplay-models-tools/blob/dev/samples/node/main.js#L10 or https://github.com/Azure/digital-twin-model-identifier#validation-regular-expressions to validate string adheres to DTMI restrictions before parsing it out.